tests/src/Interop/StringMarshalling/UTF8/UTF8TestNative.cpp

```diff
@@ -0,0 +1,319 @@
+#ifdef _WIN32	
+char* UTF16ToUTF8(wchar_t * pszTextUTF16)
+{
+    if ((pszTextUTF16 == NULL) || (*pszTextUTF16 == L'\0')) {
+        return 0;
+    }
+
+    size_t cchUTF16;
+    cchUTF16 = wcslen(pszTextUTF16) + 1;
+    int cbUTF8 = WideCharToMultiByte(CP_UTF8, 0,
+        pszTextUTF16,
+        (int)cchUTF16,
+        NULL,
+        0/* request buffer size*/,
+        NULL,
+        NULL);
+    
+    char *pszUTF8 = (char*)CoTaskMemAlloc(sizeof(char) * (cbUTF8 + 1));
+    int nc = WideCharToMultiByte(CP_UTF8, // convert to UTF-8
+        0,       //default flags 
+        pszTextUTF16, //source wide string
+        (int)cchUTF16,     // length of wide string
+        pszUTF8,      // destination buffer 
+        cbUTF8,       // destination buffer size
+        NULL,
+        NULL);
+    
+    if (!nc)
+    {
+        throw;
+    }
+
+    pszUTF8[nc] = '\0';
+    return pszUTF8;
+}
+
+wchar_t* UTF8ToUTF16(const char *utf8)
+{
+    // Special case of empty input string
+    //wszTextUTF16
+    wchar_t *wszTextUTF16 = 0;
+    if (!utf8 || !(*utf8))
+        return wszTextUTF16;
+    size_t szUtf8 = strlen(utf8);
+
+    //Get length (in wchar_t's) of resulting UTF-16 string
+    int cbUTF16 = ::MultiByteToWideChar(
+        CP_UTF8,            // convert from UTF-8
+        0,                  // default flags
+        utf8,        // source UTF-8 string
+        (int)szUtf8,      // length (in chars) of source UTF-8 string
+        NULL,               // unused - no conversion done in this step
+        0                   // request size of destination buffer, in wchar_t's
+    );
+    
+    wszTextUTF16 = (wchar_t*)(CoTaskMemAlloc((cbUTF16 + 1 )  * sizeof(wchar_t) ));
+    // Do the actual conversion from UTF-8 to UTF-16
+    int nc = ::MultiByteToWideChar(
+        CP_UTF8,            // convert from UTF-8
+        0,                  // default flags
+        utf8,        // source UTF-8 string
+        (int)szUtf8,      // length (in chars) of source UTF-8 string
+        wszTextUTF16,          // destination buffer
+        cbUTF16);  // size of destination buffer, in wchar_t's
+
+    if (!nc)
+    {
+        throw;
+    }
+    //MultiByteToWideChar do not null terminate the string when cbMultiByte is not -1
+    wszTextUTF16[nc] = '\0';
+    return wszTextUTF16;
+}
+#endif
+
+
+LPSTR build_return_string(const char* pReturn)
+{
+    char *ret = 0;
+    if (pReturn == 0 || *pReturn == 0)
+        return ret;
+
+    size_t strLength = strlen(pReturn);
+    ret = (LPSTR)(CoTaskMemAlloc(sizeof(char)* (strLength + 1)));
+    memset(ret, '\0', strLength + 1);
+    strncpy_s(ret, strLength + 1, pReturn, strLength);
+    return ret;
+}
+
+// this is the same set as in managed side , but here 
+// string need to be escaped  , still CL applied some local and 
+// end up with different byte sequence.
+
+const int NSTRINGS = 6;
+#ifdef _WIN32	
+wchar_t  *utf8Strings[] = { L"Managed",
+L"S\x00EEne kl\x00E2wen durh die wolken sint geslagen" ,
```

Comment:
There is away to store utf8 string using the u8 prefix , but then then the encoding is dependent on the locale. For example if i have char* utf8String = u8"λώσσα μου" and try to match the byte sequence from Encoding.UTF8.GetBytes ("λώσσα μου") they don't match. I haven't gone in to the details , but i don't know string literals are encoded by CL.exe . So even if i do #include or #define  still will have the same issue.